File: lib/solvers/TraceWidthSolver/isObstacleConnectedToRoute.ts
import type { ConnectivityMap } from "circuit-json-to-connectivity-map"
import type { Obstacle } from "lib/types"

type RouteConnectionIds = {
  connectionName: string
  rootConnectionName?: string
}

export const isObstacleConnectedToRoute = (
  obstacle: Obstacle,
  route: RouteConnectionIds,
  connMap?: ConnectivityMap,
) => {
  const getCanonicalConnectionId = (connectionId: string) => {
    if (typeof connMap?.getNetConnectedToId !== "function") {
      return connectionId
    }
    return connMap.getNetConnectedToId(connectionId) ?? connectionId
  }
  const routeConnectionIds = [
    route.connectionName,
    route.rootConnectionName,
  ].filter((connectionId): connectionId is string => Boolean(connectionId))

  return obstacle.connectedTo.some(
    (connectedId) =>
      routeConnectionIds.includes(connectedId) ||
      routeConnectionIds.some(
        (routeConnectionId) =>
          connMap?.areIdsConnected(routeConnectionId, connectedId) ?? false,
      ) ||
      routeConnectionIds
        .map(getCanonicalConnectionId)
        .includes(getCanonicalConnectionId(connectedId)),
  )
}

File: rust/repair/src/is_obstacle_connected_to_route.rs
use serde_json::Value;
use crate::net_utils::RepairConnectivityMap;

pub fn are_ids_connected(left: &str, right: &str, conn_map: Option<&RepairConnectivityMap>) -> bool {
    if left == right { return true; }
    let Some(map) = conn_map else { return false; };
    let Some(a) = map.id_to_net_map.get(left).filter(|s| !s.is_empty()) else { return false; };
    let Some(b) = map.id_to_net_map.get(right).filter(|s| !s.is_empty()) else { return false; };
    a == b || b == left
}

pub fn is_obstacle_connected_to_route(obstacle: &Value, route: &Value, conn_map: Option<&RepairConnectivityMap>) -> bool {
    let name = route["connectionName"].as_str().expect("Route connectionName");
    let root = route["rootConnectionName"].as_str();
    obstacle["connectedTo"].as_array().expect("Obstacle connectedTo").iter().filter_map(Value::as_str).any(|id| {
        id == name || Some(id) == root || (conn_map.is_some() && (are_ids_connected(name, id, conn_map) || root.is_some_and(|root| are_ids_connected(root, id, conn_map))))
    })
}
